doc: remove pessimistic gas from nearcore dev book (#13514)
near/nearcore: docs/architecture/how/gas.md

2. [Gas Price](#gas-price):
    - [Block-Level Gas Price](#block-level-gas-price): How the block-level gas price is determined.
    - [Pessimistic Gas Price](#pessimistic-gas-price): How worst-case gas pricing is estimated.

>     - [Gas Refund Tax](#gas-refund-tax): The cost paid for a receipt.

Reviewer: kind of nitpicky but I would not call it a "tax" or "penalty", "Gas refund fee" sounds best to me.

Author: Hm, the NEP authors specifically called it a penalty. Empathizing that it's not just a regular fee covering chunk space but rather a disincentive to not attach too much gas.

Personally, I don't like "penalty" too much, either. But I've already used it in the code implementing all of this, including the config options here: https://github.com/near/nearcore/blob/f381b9dbf0c8381892c821341ed6fe62a8125502/core/parameters/res/runtime_configs/78.yaml#L10-L20

We can still go with "gas refund fee" for the documentation here. Then I can also use this terminology for the contract developer docs and the nomicon, keeping it persistent. But I think in the nearcore development guide, I should at least mention in a side comment that we call it "penalty" in the code and why.

A signer pays all the gas required for a transaction upfront. However, there is
no explicit act of buying gas. Instead, the fee is subtracted directly in NEAR
tokens from the balance of the signer's account. The fee is calculated as `gas

> amount` * `gas price`.

Reviewer: Can we add something like:
Gas amount for actions included in a `SignedTransaction` are all fixed, except for function calls where the user needs to specify the amount.

possibly with a link to here: https://docs.near.org/protocol/gas#cost-for-common-actions

> If the account has insufficient balance to pay for this pessimistic pricing, it

Reviewer: ```suggestion
If the account has insufficient balance to pay for this, it
```

Already we can see a fundamental problem: Gas is bought once at the beginning
but the gas price may change during execution. To solve this incompatibility,
the protocol used to calculate a pessimistic gas price for the initial purchase.
Later on, the delta between real and pessimistic gas prices would be refunded at
the end of every receipt execution.

Generating a refund receipts for every executed function call has become a
non-trivial overhead, limiting total throughput. Therefore, with the adoption of
[NEP-536](https://github.com/near/NEPs/pull/536) the model was changed.

> Today, the protocol simply ignores gas price changes between time of purchase

Reviewer 2: nit: maybe mention protocol v78 here as well? (e.g. v78 and onward, ...)

### Pessimistic Gas Price

The pessimistic gas price features was removed with protocol version 78 and
[NEP-536](https://github.com/near/NEPs/pull/536).

### Gas Refund Tax

After executing the transaction, there might be unspent gas left. For function
calls, this is the normal case, since attaching exactly the right amount of gas
is tricky. Additionally, in case of receipt failure, some actions that did not
start execution will have the execution gas unspent. This gas is converted back
to NEAR tokens and sent as a refund transfer to the original signer.

Before protocol version 78, the full gas amount would be refunded and the refund
transfer action was executed for free. Since version 78, there is a gas refund

> tax, acting as a penalty for attaching too much gas. The tax is 5% of unspent
> gas but at least 1 Tgas. This often removes the need to send a refund, avoiding
> additional load on the network.

Reviewer: ```suggestion
fee. The network charges a fee of 5% of unspent gas or a minimum of 1 Tgas.
This often removes the need to send a refund, which avoids additional load on the network, and it compensates for the load of the refund receipt when it needs to be issued. This is intended to discourage users from attaching more gas than needed to their transactions.
```

Reviewer: not sure if this is an improvement, mainly trying to avoid "tax" here.

Technically, the refund transfer is still executed for free, since the gas price
is set to 0. This keeps it in line with balance refunds. However, when the gas
refund is produced, at least 1 Tgas has been burnt on receipt execution, which
more than covers for a transfer.

> Finally, we can have a complete diagram, including the gas gas penalty.

Reviewer: ```suggestion
Finally, we can have a complete diagram, including the gas refund penalty.
```

to NEAR tokens and sent as a refund transfer to the original signer.

Before protocol version 78, the full gas amount would be refunded and the refund

> transfer action was executed for free. Since version 78, thhe network charges a

Reviewer: ```suggestion
transfer action was executed for free. Since version 78, the network charges a
```
